### PR TITLE
Modularize iptables manager

### PR DESCRIPTION
**Pull request by pippolo84**
Modularize the iptables configuration manager in a cell.
Direct dependency to the DaemonConfig global object is avoided through a SharedConfig privately provided to the cell. This config gets the values from all the referenced DaemonConfig helpers at startup, in order to be used by the new cell.

Also, cells depending on iptables (like the node manager one) now takes a reference to it through dependency injection instead of directly calling side effectful exported functions.

~Depends on https://github.com/cilium/cilium/pull/28713~

### CODE REVIEW

File: pkg/modules/cell.go
Comment (joamaki) [inline pkg/modules/cell.go]: No need to repeat the module id in the title, e.g. you could just have "Loads kernel modules required by Cilium".
This package also should be relocated to `pkg/datapath/linux/modules`.

File: pkg/modules/modules.go
Comment (joamaki) [inline pkg/modules/modules.go]: bit of code golf:
```
func (hive.HookContext) (err error) {
  m.modulesList, err = listModules()
  return
}
```
or maybe even better would be just:

```
func (m *Manager) Start(hive.HookContext) (err error) {
  m.moduleList, err = listModules()
  return
}

func newManager(...) ... {
  ...
  lc.Append(hive.Hook{OnStart: m.Start})
}
```

File: pkg/modules/cell.go
Comment (joamaki) [inline pkg/modules/cell.go]: Maybe `kernel-module-loader` would be more informative?

File: pkg/datapath/iptables/cell.go
Comment (joamaki) [inline pkg/datapath/iptables/cell.go]: I would just call this `iptables` as `-manager` doesn't really add anything.

File: pkg/datapath/iptables/cell.go
Comment (joamaki) [inline pkg/datapath/iptables/cell.go]: No need to repeat module id in title

File: pkg/datapath/cells.go
Comment (joamaki) [inline pkg/datapath/cells.go]: I would move this into `iptables.Cell` and keep things fairly tight here.

File: pkg/datapath/iptables/iptables.go
Comment (joamaki) [inline pkg/datapath/iptables/iptables.go]: Don't fatal, but return an error instead.

File: pkg/datapath/iptables/iptables.go
Comment (joamaki) [inline pkg/datapath/iptables/iptables.go]: The error is now dropped, I would wrap it instead with fmt.Errorf and `%w`.

**Review by joamaki — CHANGES_REQUESTED** [review]

File: pkg/modules/cell.go
Comment (pippolo84) [reply pkg/modules/cell.go]: Fixed in https://github.com/cilium/cilium/pull/28713 as well as for the other comments regarding modules manager.

**Comment by pippolo84** [comment]
I've addressed all the feedbacks (the ones regarding modules manager have been pushed in the specific PR as well) but I'm planning to rework this quite a lot, so still not ready for review yet.

File: pkg/datapath/iptables/cell.go
Comment (dylandreimerink) [inline pkg/datapath/iptables/cell.go]: This is essentially just a layer around the daemonConfig, I don't see the need to expose this as a separate type in hive. I understand it makes sense internally for the purposes of mocking though

**Review by dylandreimerink — CHANGES_REQUESTED** [review]

File: pkg/datapath/iptables/cell.go
Comment (pippolo84) [reply pkg/datapath/iptables/cell.go]: Yep, main reason to do that is to avoid depending on the big global config object `DaemonConfig`. But it's better to provide it only privately. I'll fix this, thanks :+1:

File: pkg/datapath/iptables/cell.go
Comment (mhofstetter) [inline pkg/datapath/iptables/cell.go]: changing this to a Hive config no longer parses the passed string as space-separated string. Instead we have to adapt the Helm Chart to pass a comma-separated string -> https://github.com/cilium/cilium/blob/main/install/kubernetes/cilium/templates/cilium-configmap.yaml#L926

File: pkg/datapath/iptables/cell.go
Comment (mhofstetter) [inline pkg/datapath/iptables/cell.go]: i'd prefer having these const values inlined. It's still possible to keep the comments above the corresponding flag definitions in `Flags(...)`

File: pkg/datapath/iptables/iptables.go
Comment (dylandreimerink) [inline pkg/datapath/iptables/iptables.go]: If we fail to set the sysctl values, it does not necessity mean that Cilium will not function (the correct values may be set but we don't have write access), so I would convert this to a warning like the check below.

**Review by dylandreimerink — APPROVED** [review]
LGTM. Just one nit, but given its copied code will not block on that

File: pkg/datapath/cells.go
Comment (joamaki) [inline pkg/datapath/cells.go]: Maybe better as something like "Manages Cilium-specific iptables rules"?

File: pkg/datapath/iptables/cell.go
Comment (joamaki) [reply pkg/datapath/iptables/cell.go]: Yep likewise. I would move the doc comments on to the struct fields.

**Review by joamaki — APPROVED** [review]

File: pkg/datapath/iptables/cell.go
Comment (pippolo84) [reply pkg/datapath/iptables/cell.go]: I've put this on hold to wait for the final decision about how to approach the StringSlice issue: https://github.com/cilium/cilium/pull/28919#issuecomment-1790405427

Comment (pippolo84) [reply pkg/datapath/iptables/cell.go]: I'll take care of this in a follow-up PR as soon as we reach consensus.

**Review by mhofstetter — APPROVED** [review]
Thanks - LGTM!

**Comment by pippolo84** [comment]
Force-pushed to fix controlplane tests: `... missing type: *iptables.Manager ...`

**Comment by pippolo84** [comment]
Conformance Cluster Mesh failure tracked [here](https://github.com/cilium/cilium/issues/24990#issuecomment-1801895280), rerunning.

**Pull request merged**
